File: array/ts/squares_of_a_sorted_array.ts
// 977. 有序数组的平方 (Squares of a Sorted Array)
// 双指针：平方后最大值只可能来自首尾两端，从结果数组末尾往前填。
function sortedSquares(nums: number[]): number[] {
    const n = nums.length;
    const res = new Array<number>(n);

    let i = 0, j = n - 1;
    for (let k = n - 1; k >= 0; k--) {
        const a = nums[i] * nums[i];
        const b = nums[j] * nums[j];
        if (a > b) {
            res[k] = a;
            i++;
        } else {
            res[k] = b;
            j--;
        }
    }
    return res;
}

File: array/python/squares_of_a_sorted_array.py
# 977. 有序数组的平方 (Squares of a Sorted Array)
# 双指针：平方后最大值只可能来自首尾两端，从结果数组末尾往前填。
from typing import List

class Solution:
    def sortedSquares(self, nums: List[int]) -> List[int]:
        n = len(nums)
        res = [0] * n

        i, j = 0, n - 1
        for k in range(n - 1, -1, -1):
            a, b = nums[i] * nums[i], nums[j] * nums[j]
            if a > b:
                res[k] = a
                i += 1
            else:
                res[k] = b
                j -= 1

        return res
